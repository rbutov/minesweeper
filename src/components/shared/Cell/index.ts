enum CellTypes {
  CLOSED = 'closed',
  FLAG = 'flag',
  MINE = 'mine',
  MINE_RED = 'mine_red',
  MINE_WRONG = 'mine_wrong',
  TYPE_0 = 'type_0',
  TYPE_1 = 'type_1',
  TYPE_2 = 'type_2',
  TYPE_3 = 'type_3',
  TYPE_4 = 'type_4',
}

type CellType = `${CellTypes}`;

interface CellProps {
  type: CellType;
  row: number;
  col: number;
}

import closed from 'assets/cell/closed.svg';
import flag from 'assets/cell/flag.svg';
import mine from 'assets/cell/mine.svg';
import mine_red from 'assets/cell/mine_red.svg';
import mine_wrong from 'assets/cell/mine_wrong.svg';
import type0 from 'assets/cell/type0.svg';
import type1 from 'assets/cell/type1.svg';
import type2 from 'assets/cell/type2.svg';
import type3 from 'assets/cell/type3.svg';
import type4 from 'assets/cell/type4.svg';

const CellAsserts = {
  [CellTypes.CLOSED]: closed,
  [CellTypes.FLAG]: flag,
  [CellTypes.MINE]: mine,
  [CellTypes.MINE_RED]: mine_red,
  [CellTypes.MINE_WRONG]: mine_wrong,
  [CellTypes.TYPE_0]: type0,
  [CellTypes.TYPE_1]: type1,
  [CellTypes.TYPE_2]: type2,
  [CellTypes.TYPE_3]: type3,
  [CellTypes.TYPE_4]: type4,
};

export * from './Cell';
export { CellTypes, CellAsserts };
export type { CellProps, CellType };
