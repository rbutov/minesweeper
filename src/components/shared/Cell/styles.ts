import styled from 'styled-components';
import { CellAsserts, CellType, CellTypes } from '.';

const CellContainer = styled.div<{ type: CellType; failedMineKey: string | null }>`
  width: 24px;
  height: 24px;
  font-size: 10px;
  line-height: 23px;
  float: left;
  background-size: 100%;
  font-weight: 700;
  text-align: center;

  ${({ type }) => `background-image: url(${CellAsserts[type]})`};

  :active {
    ${({ type, failedMineKey }) =>
      type === CellTypes.CLOSED && !failedMineKey
        ? `background-image: url(${CellAsserts[CellTypes.TYPE_0]})`
        : ''};
  }
`;

export { CellContainer };
