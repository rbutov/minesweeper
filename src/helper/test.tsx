import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import { ReactWrapper, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { confStore } from 'store';

type Wrapper<P = unknown, S = unknown> = ShallowWrapper<P, S> | ReactWrapper<P, S>;
type MountFnType<P, S> = (
  node: React.ReactElement<P>,
  options?: Record<string, unknown>
) => Wrapper<P, S>;

const wrapAppComponent = <P, S>(
  renderer: MountFnType<P, S>,
  element: JSX.Element,
  other?: {
    store?: Store;
  }
): Wrapper<P, S> => {
  const coreStore = confStore();

  return renderer(<Provider store={other?.store ? other.store : coreStore}>{element}</Provider>);
};

const waitForComponentToPaint = async <P, S>(
  wrapper: Wrapper<P, S>,
  other?: { delay?: number; props?: Pick<P, keyof P> }
) => {
  await act(async () => {
    await runAllPromises();
    await new Promise((resolve) => setTimeout(resolve, other?.delay ? other.delay : 0));
    wrapper.setProps({ ...other?.props });
    wrapper.update();
  });
};

const runAllPromises = async () => {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, 0);
  });
};

export { wrapAppComponent, waitForComponentToPaint };
export type { Wrapper };
