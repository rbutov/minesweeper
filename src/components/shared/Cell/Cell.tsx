import React from 'react';
import { useAppDispatch } from 'hooks/useAppDispatch';
import { revealKey, toggleFlag } from 'actions/fieldActions';
import { CellContainer } from './styles';
import { CellProps } from '.';

const Cell = ({ type, row, col, failedMineKey }: CellProps) => {
  const dispatch = useAppDispatch();

  const handleMouseDown = (ev: React.MouseEvent<HTMLElement>) => {
    ev.preventDefault();

    if (ev.button === 2) {
      dispatch(toggleFlag({ row, col }));
      return;
    }
  };

  const handleClick = (ev: React.MouseEvent<HTMLElement>) => {
    ev.preventDefault();
    dispatch(revealKey({ row, col }));
  };

  return (
    <CellContainer
      type={type}
      failedMineKey={failedMineKey}
      onMouseDown={handleMouseDown}
      onClick={handleClick}
    />
  );
};

export { Cell };
