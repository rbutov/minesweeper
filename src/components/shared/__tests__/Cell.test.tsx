import React from 'react';
import { wrapAppComponent } from 'helpers/test';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Cell, CellTypes } from '../Cell';

describe('<Cell />', () => {
  it('_render', () => {
    const wrapper = wrapAppComponent(
      mount,
      <Cell row={0} col={0} failedMineKey={'0'} type={CellTypes.CLOSED} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('_mine', () => {
    const wrapper = wrapAppComponent(
      mount,
      <Cell row={0} col={0} failedMineKey={'0'} type={CellTypes.MINE} />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
