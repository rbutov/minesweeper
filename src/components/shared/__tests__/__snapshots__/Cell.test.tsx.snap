// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> _mine 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo()
    col={0}
    failedMineKey="0"
    row={0}
    type="mine"
  >
    <styled.div
      data-testid="cell"
      failedMineKey="0"
      onClick={[Function]}
      onMouseDown={[Function]}
      type="mine"
    >
      <div
        className="sc-bdvvtL guyQzE"
        data-testid="cell"
        onClick={[Function]}
        onMouseDown={[Function]}
        type="mine"
      />
    </styled.div>
  </Memo()>
</Provider>
`;

exports[`<Cell /> _render 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Memo()
    col={0}
    failedMineKey="0"
    row={0}
    type="closed"
  >
    <styled.div
      data-testid="cell"
      failedMineKey="0"
      onClick={[Function]}
      onMouseDown={[Function]}
      type="closed"
    >
      <div
        className="sc-bdvvtL htSWkN"
        data-testid="cell"
        onClick={[Function]}
        onMouseDown={[Function]}
        type="closed"
      />
    </styled.div>
  </Memo()>
</Provider>
`;
